Refactor the Extensions settings page

Nokel81 opened the pull request:
- Simplify the install logic by refactoring out the multi-install logic

- Revamp the `ExtensionInstallStateStore` to more strictly track the lifetime of an install or uninstall request

- Fix the install of an already installed extension just hanging visually

- Display a spinner more often for more visual feedback

Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes #2116

Nokel81 committed:
Refactor the Extensions settings page

- Simplify the install logic by refactoring out the multi-install logic

- Revamp the ExtensionInstallStateStore to more strictly track the
  lifetime of an install or uninstall request

- Fix the install of an already installed extension just hanging
  visually

- Display a spinner more often for more visual feedback

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on src/renderer/components/+extensions/extension-install.store.ts:
```diff
@@ -1,13 +1,187 @@
+const preInstallIds = observable.set<string>();
+
+export class ExtensionInstallationStateStore {
+  @action static reset() {
```
I see only `reset` being used .. are other methods going to be used in some other PR?

Nokel81 replied:
Do you check `src/renderer/components/+extensions/extensions.tsx`? It is by default not shown because it is considered "large"

jakolehm replied:
Missed that, now it makes more sense 🙂

Nokel81 committed:
Fix some edge cases

- Clear the loading state in more places
- Add loading animation when an extension is discovered by main

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix unit test

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix getting error message for displaying to the user

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
some fixes to unit tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix unit tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
remove flaky test

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Add support for installing extensions by name (#2435)

* Add support for installing extensions by name

- Optionally specify the version, defaults to latest non-prerelease
  version

- Usable via the protocol handlers, but requires user confirmation to
  actually install

Signed-off-by: Sebastian Malton <sebastian@malton.name>

* resolve pr comments

Signed-off-by: Sebastian Malton <sebastian@malton.name>

* add confirm dialog message

Signed-off-by: Sebastian Malton <sebastian@malton.name>

* display error when registry returns not found as JSON

Signed-off-by: Sebastian Malton <sebastian@malton.name>

* fix unit test

Signed-off-by: Sebastian Malton <sebastian@malton.name>

* fix registry error detection

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/extensions/extension-discovery.ts:
```diff
@@ -190,8 +193,9 @@ export class ExtensionDiscovery {
     // Check that the removed path is directly under this.localFolderPath
```
Is the comment on line 189 true anymore? (github won't let me comment on that line, this is the nearest I can get)
```
// filePath is the non-symlinked path to the extension folder
```
On line 149 `handleWatchUnlinkDir` is added as a handler for `unlink` (watching `"<extensionSymLink>"`)

jim-docker commented on src/extensions/extension-discovery.ts:
```diff
@@ -272,26 +267,26 @@ export class ExtensionDiscovery {
+      return await this.ensureExtensions();
+    } finally {
+      this.isLoaded = true;
+    }
```
Did you mean to change behaviour here? (previously `this.isLoaded` was not set to `true` if `this.ensureExtensions()` threw an error.

jim-docker commented on src/extensions/extension-loader.ts:
```diff
@@ -290,28 +288,20 @@ export class ExtensionLoader {
-      if (extEntrypoint !== "") {
-        if (!fs.existsSync(extEntrypoint)) {
-          console.log(`${logModule}: entrypoint ${extEntrypoint} not found, skipping ...`);
+    const extensionEntryPointAbsolutePath = path.resolve(path.join(path.dirname(extension.manifestPath), extensionEntryPointRelativePath));
```
these path functions can all throw a `TypeError` so why not keep this statement in the `try` block?

Nokel81 replied:
They can throw but only if it is not a string.

jim-docker commented on src/renderer/components/+extensions/__tests__/extensions.test.tsx:
```diff
@@ -46,84 +47,55 @@ jest.mock("../../../../extensions/extension-loader", () => ({
+    (fse.unlink as jest.MockedFunction<typeof fse.unlink>).mockReturnValue(Promise.resolve() as any);
 
-    fireEvent.change(screen.getByPlaceholderText("Path or URL to an extension package", {
+    fireEvent.change(res.getByPlaceholderText("Path or URL to an extension package", {
```
The placeholder text has changed but I guess this still works because it's not looking for an exact match?

Nokel81 replied:
Probably

jim-docker commented on src/renderer/components/+extensions/extension-install.store.ts:
```diff
@@ -1,13 +1,218 @@
+export enum ExtensionInstallationState {
+  INSTALLING = "installing",
+  UNINSTALLING = "uninstalling",
+  IDLE = "IDLE",
```
```suggestion
  IDLE = "idle",
```

jim-docker commented on src/renderer/components/+extensions/extensions.tsx:
```diff
@@ -35,458 +57,457 @@ interface InstallRequestPreloaded extends InstallRequest {
+    parseJson: true,
+  });
+
+  if (!manifest.lens && !manifest.renderer) {
```
why `manifest.lens` and not `manifest.main`?

Nokel81 replied:
My mistake

jim-docker commented on src/renderer/components/+extensions/extensions.tsx:
```diff
@@ -35,458 +57,457 @@ interface InstallRequestPreloaded extends InstallRequest {
+      <div className="flex column gaps">
+        <b>Extension Install Collision:</b>
+        <p>The <em>{name}</em> extension is currently {curState.toLowerCase()}.</p>
+        <p>Will not procede with this current install request.</p>
```
```suggestion
        <p>Will not proceed with this current install request.</p>
```

jim-docker commented on src/renderer/components/+extensions/extensions.tsx:
```diff
@@ -35,458 +57,457 @@ interface InstallRequestPreloaded extends InstallRequest {
+          if (await uninstallExtension(validatedRequest.id, validatedRequest.manifest)) {
+            await unpackExtension(validatedRequest, dispose);
+          } else {
+            dispose();
```
Does the `validatedRequest.tempFile` need to be removed at this point?

Nokel81 replied:
Yes

jim-docker commented on src/renderer/components/+extensions/extensions.tsx:
```diff
@@ -35,458 +57,457 @@ interface InstallRequestPreloaded extends InstallRequest {
+      </div>,
+      {
+        onClose() {
+          dispose();
```
and here?

jim-docker commented on src/renderer/components/+extensions/extensions.tsx:
```diff
@@ -35,458 +57,457 @@ interface InstallRequestPreloaded extends InstallRequest {
+    logger.info(`[EXTENSION-INSTALL]: installation has failed: ${message}`, { error, installPath });
+    Notifications.error(<p>Installation has failed: <b>{message}</b></p>);
+  } finally {
+    disposer?.();
```
doesn't `requestInstall()` call this in the case where `disposer` is not null?

Nokel81 replied:
The disposer that I created can be called multiple times safely.

jim-docker replied:
Could you have added the removal of `validatedRequest.tempFile` to the disposer?

Nokel81 replied:
I tried that and I though it was too complicated. I tried it again and it isn't.

Nokel81 committed:
respond to review comments

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
better use of disposers

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/renderer/components/+extensions/extensions.tsx:
```diff
@@ -366,15 +364,11 @@ async function attemptInstall(request: InstallRequest, d?: Disposer): Promise<vo
-          dispose();
-          fse.unlink(validatedRequest.tempFile).catch(noop);
-        }
+        onClose: dispose,
```
much cleaner!

jakolehm reviewed (approved):
LGTM
